docs/CONTRIBUTING.md

```diff
@@ -61,6 +61,24 @@ are squashed against the correct commit if necessary.
 
 ## Style Guide
 
+### Logging
+
+When possible, use log levels according to the following guidelines (in
+decreasing order of severity):
+
+- `error`: there is definitely a logic bug somewhere in the code, and someone
+  should look into it; it should not be possible to trigger an `error` log just
+  by manipulating inputs / configuration details
```

Comment:
I wonder if that goal is realistic. It seems like there may be a few cases in which an error should be thrown based on user input, following the model that David outlined, which imho corresponds to how most devs see errors/warnings. 

Maybe I'm biased coming from a dynamically typed lang in which runtime exceptions are common, but it seems as if there are a few cases of invalid input that we cannot prevent with types, such as invalid URI like @rbehjati mentioned.

Reply:
Eg in REST land you sort of a separation there

The 400 code basically means "you provided invalid input, and it's serious so we can't proceed" (error that can happen)

The 500 code means "something internally fucked up" aka there is definitely a logic bug somewhere in the code, and someone should look into it.

Reply:
My naive two cents of moving this over our codebase in a way that seems intuitive, to me, a newcomer:

panic: should never happen, def should be looked into
error: something serious went wrong, cannot proceed, possibly due to config/input
warning: suspicious, but can proceed, ideally should only be triggered by suspicious/invalid input

Reply:
I think we agree on the distinction between 4XX and 5XX errors, the question here is really whether the boundary between them is the logic in the Runtime, or does it extend also to the logic of the particular Oak Application that is running on it (and if so, all of it, or just interactions via the ABI).

Reply:
Thanks for clarifying, I feel like perhaps the difference between runtime vs application logs wasn't as clear to me.

If our we think it's possible to create a world where an application can never leave the runtime in a state where it can't reasonably carry on as expected, then yes that classification of feels reasonable. :)